File: src/backEnd/model/ScanPreset.py
"""
扫描配置预设模型
用于存储常用配置、默认配置、历史配置
配置字段与sqlmap的optiondict.py保持一致
"""
from pydantic import BaseModel, Field
from typing import Optional, Dict, Any, List
from datetime import datetime
from enum import Enum


class PresetType(str, Enum):
    """预设类型"""
    DEFAULT = "default"      # 默认配置
    PRESET = "preset"        # 常用配置
    HISTORY = "history"      # 历史配置


class ScanOptions(BaseModel):
    """
    扫描选项模型
    字段名与sqlmap的optiondict.py保持一致
    """
    # ==================== Detection 检测选项 ====================
    level: int = Field(default=1, ge=1, le=5, description="检测级别 (1-5)")
    risk: int = Field(default=1, ge=1, le=3, description="风险级别 (1-3)")
    string: Optional[str] = Field(default=None, description="页面匹配字符串")
    notString: Optional[str] = Field(default=None, description="页面不匹配字符串")
    regexp: Optional[str] = Field(default=None, description="正则匹配")
    code: Optional[int] = Field(default=None, description="HTTP响应码")
    smart: bool = Field(default=False, description="智能检测")
    textOnly: bool = Field(default=False, description="仅文本比较")
    titles: bool = Field(default=False, description="基于标题比较")
    
    # ==================== Injection 注入选项 ====================
    testParameter: Optional[str] = Field(default=None, description="指定测试参数 (-p)")
    skip: Optional[str] = Field(default=None, description="跳过参数")
    skipStatic: bool = Field(default=False, description="跳过静态参数")
    paramExclude: Optional[str] = Field(default=None, description="排除参数")
    dbms: Optional[str] = Field(default=None, description="数据库类型")
    os: Optional[str] = Field(default=None, description="操作系统")
    prefix: Optional[str] = Field(default=None, description="注入前缀")
    suffix: Optional[str] = Field(default=None, description="注入后缀")
    tamper: Optional[str] = Field(default=None, description="篡改脚本")
    
    # ==================== Techniques 技术选项 ====================
    technique: str = Field(default="BEUSTQ", description="注入技术 (BEUSTQ)")
    timeSec: int = Field(default=5, ge=1, le=60, description="时间盲注延迟(秒)")
    
    # ==================== Request 请求选项 ====================
    timeout: int = Field(default=30, ge=1, description="请求超时(秒)")
    retries: int = Field(default=3, ge=0, description="重试次数")
    delay: float = Field(default=0, ge=0, description="请求延迟(秒)")
    randomAgent: bool = Field(default=False, description="随机User-Agent")
    proxy: Optional[str] = Field(default=None, description="代理")
    tor: bool = Field(default=False, description="使用Tor")
    
    # ==================== Optimization 优化选项 ====================
    optimize: bool = Field(default=False, description="使用所有优化选项")
    predictOutput: bool = Field(default=False, description="预测输出")
    keepAlive: bool = Field(default=False, description="保持连接")
    nullConnection: bool = Field(default=False, description="空连接")
    threads: int = Field(default=1, ge=1, le=10, description="线程数")
    
    # ==================== Enumeration 枚举选项 ====================
    getBanner: bool = Field(default=False, description="获取Banner")
    getCurrentUser: bool = Field(default=False, description="获取当前用户")
    getCurrentDb: bool = Field(default=False, description="获取当前数据库")
    getHostname: bool = Field(default=False, description="获取主机名")
    isDba: bool = Field(default=False, description="是否DBA")
    getUsers: bool = Field(default=False, description="获取所有用户")
    getPasswordHashes: bool = Field(default=False, description="获取密码哈希")
    getPrivileges: bool = Field(default=False, description="获取权限")
    getRoles: bool = Field(default=False, description="获取角色")
    getDbs: bool = Field(default=False, description="获取所有数据库")
    getTables: bool = Field(default=False, description="获取所有表")
    getColumns: bool = Field(default=False, description="获取所有列")
    dumpTable: bool = Field(default=False, description="导出表")
    dumpAll: bool = Field(default=False, description="导出所有")
    db: Optional[str] = Field(default=None, description="指定数据库")
    tbl: Optional[str] = Field(default=None, description="指定表")
    col: Optional[str] = Field(default=None, description="指定列")
    
    # ==================== General 通用选项 ====================
    batch: bool = Field(default=True, description="非交互模式")
    forms: bool = Field(default=False, description="解析表单")
    crawlDepth: int = Field(default=0, ge=0, description="爬取深度(0=禁用)")
    flushSession: bool = Field(default=False, description="刷新会话")
    freshQueries: bool = Field(default=False, description="刷新查询")
    verbose: int = Field(default=1, ge=0, le=6, description="详细级别 (0-6)")
    
    def to_dict(self) -> Dict[str, Any]:
        """转换为字典，只包含非默认值"""
        result = {}
        default_model = ScanOptions()
        for field_name, field_value in self:
            default_value = getattr(default_model, field_name)
            if field_value != default_value and field_value is not None:
                result[field_name] = field_value
        return result
    
    def to_full_dict(self) -> Dict[str, Any]:
        """转换为完整字典"""
        return self.model_dump()


class ScanPreset(BaseModel):
    """扫描配置预设"""
    id: Optional[int] = Field(default=None, description="预设ID")
    name: str = Field(..., min_length=1, max_length=100, description="预设名称")
    description: Optional[str] = Field(default=None, max_length=500, description="预设描述")
    preset_type: PresetType = Field(default=PresetType.PRESET, description="预设类型")
    options: ScanOptions = Field(default_factory=ScanOptions, description="扫描选项")
    parameter_string: Optional[str] = Field(default=None, description="命令行参数字符串(与BurpSuite兼容)")
    is_active: bool = Field(default=True, description="是否启用")
    created_at: Optional[datetime] = Field(default=None, description="创建时间")
    updated_at: Optional[datetime] = Field(default=None, description="更新时间")
    last_used_at: Optional[datetime] = Field(default=None, description="最后使用时间")
    use_count: int = Field(default=0, ge=0, description="使用次数")
    
    model_config = {
        "use_enum_values": True,
        "json_encoders": {
            datetime: lambda v: v.isoformat() if v else None
        }
    }
    
    def model_dump(self, **kwargs):
        """Override model_dump to convert datetime to string"""
        data = super().model_dump(**kwargs)
        # Convert datetime fields to ISO format strings
        for field in ['created_at', 'updated_at', 'last_used_at']:
            if field in data and data[field] is not None:
                if isinstance(data[field], datetime):
                    data[field] = data[field].isoformat()
        return data


class ScanPresetCreate(BaseModel):
    """创建扫描配置预设请求"""
    name: str = Field(..., min_length=1, max_length=100, description="预设名称")
    description: Optional[str] = Field(default=None, max_length=500, description="预设描述")
    preset_type: PresetType = Field(default=PresetType.PRESET, description="预设类型")
    options: Dict[str, Any] = Field(default_factory=dict, description="扫描选项")
    parameter_string: Optional[str] = Field(default=None, description="命令行参数字符串")
    
    class Config:
        use_enum_values = True


class ScanPresetUpdate(BaseModel):
    """更新扫描配置预设请求"""
    name: Optional[str] = Field(default=None, min_length=1, max_length=100, description="预设名称")
    description: Optional[str] = Field(default=None, max_length=500, description="预设描述")
    options: Optional[Dict[str, Any]] = Field(default=None, description="扫描选项")
    parameter_string: Optional[str] = Field(default=None, description="命令行参数字符串")
    is_active: Optional[bool] = Field(default=None, description="是否启用")
    
    class Config:
        use_enum_values = True


class ScanPresetListResponse(BaseModel):
    """扫描配置预设列表响应"""
    presets: List[ScanPreset] = Field(default_factory=list, description="预设列表")
    total: int = Field(default=0, description="总数")
    default_preset: Optional[ScanPreset] = Field(default=None, description="默认配置")


# ==================== 预定义配置 ====================

def create_default_preset() -> ScanPreset:
    """创建默认配置"""
    return ScanPreset(
        id=0,
        name="默认配置",
        description="系统默认扫描配置",
        preset_type=PresetType.DEFAULT,
        options=ScanOptions(),
        is_active=True,
        created_at=datetime.now(),
        updated_at=datetime.now()
    )


def create_quick_scan_preset() -> ScanPreset:
    """创建快速扫描配置"""
    return ScanPreset(
        name="快速扫描",
        description="快速扫描 - 仅基础检测",
        preset_type=PresetType.PRESET,
        options=ScanOptions(
            level=1,
            risk=1,
            technique="B",
            batch=True
        )
    )


def create_deep_scan_preset() -> ScanPreset:
    """创建深度扫描配置"""
    return ScanPreset(
        name="深度扫描",
        description="深度扫描 - 全面检测",
        preset_type=PresetType.PRESET,
        options=ScanOptions(
            level=5,
            risk=3,
            technique="BEUSTQ",
            batch=True,
            threads=3
        )
    )


def create_safe_scan_preset() -> ScanPreset:
    """创建安全扫描配置"""
    return ScanPreset(
        name="安全扫描",
        description="安全扫描 - 低风险检测",
        preset_type=PresetType.PRESET,
        options=ScanOptions(
            level=3,
            risk=1,
            technique="BEU",
            batch=True,
            delay=1
        )
    )

File: src/frontEnd/src/types/scanPreset.ts
/**
 * 扫描配置预设类型定义
 */

// 预设类型
export type PresetType = 'default' | 'preset' | 'history'

// 扫描选项
export interface ScanOptions {
  // Detection 检测选项
  level?: number          // 检测级别 (1-5)
  risk?: number           // 风险级别 (1-3)
  string?: string         // 页面匹配字符串
  notString?: string      // 页面不匹配字符串
  regexp?: string         // 正则匹配
  code?: number           // HTTP响应码
  smart?: boolean         // 智能检测
  textOnly?: boolean      // 仅文本比较
  titles?: boolean        // 基于标题比较
  
  // Injection 注入选项
  testParameter?: string  // 指定测试参数 (-p)
  skip?: string           // 跳过参数
  skipStatic?: boolean    // 跳过静态参数
  paramExclude?: string   // 排除参数
  dbms?: string           // 数据库类型
  os?: string             // 操作系统
  prefix?: string         // 注入前缀
  suffix?: string         // 注入后缀
  tamper?: string         // 篡改脚本
  
  // Techniques 技术选项
  technique?: string      // 注入技术 (BEUSTQ)
  timeSec?: number        // 时间盲注延迟(秒)
  
  // Request 请求选项
  timeout?: number        // 请求超时(秒)
  retries?: number        // 重试次数
  delay?: number          // 请求延迟(秒)
  randomAgent?: boolean   // 随机User-Agent
  proxy?: string          // 代理
  tor?: boolean           // 使用Tor
  
  // Optimization 优化选项
  optimize?: boolean      // 使用所有优化选项
  predictOutput?: boolean // 预测输出
  keepAlive?: boolean     // 保持连接
  nullConnection?: boolean // 空连接
  threads?: number        // 线程数
  
  // Enumeration 枚举选项
  getBanner?: boolean     // 获取Banner
  getCurrentUser?: boolean // 获取当前用户
  getCurrentDb?: boolean  // 获取当前数据库
  getHostname?: boolean   // 获取主机名
  isDba?: boolean         // 是否DBA
  getUsers?: boolean      // 获取所有用户
  getPasswordHashes?: boolean // 获取密码哈希
  getPrivileges?: boolean // 获取权限
  getRoles?: boolean      // 获取角色
  getDbs?: boolean        // 获取所有数据库
  getTables?: boolean     // 获取所有表
  getColumns?: boolean    // 获取所有列
  dumpTable?: boolean     // 导出表
  dumpAll?: boolean       // 导出所有
  db?: string             // 指定数据库
  tbl?: string            // 指定表
  col?: string            // 指定列
  
  // General 通用选项
  batch?: boolean         // 非交互模式
  forms?: boolean         // 解析表单
  crawlDepth?: number     // 爬取深度(0=禁用)
  flushSession?: boolean  // 刷新会话
  freshQueries?: boolean  // 刷新查询
  verbose?: number        // 详细级别 (0-6)
  
  // 其他选项
  [key: string]: any
}

// 扫描配置预设
export interface ScanPreset {
  id?: number
  name: string
  description?: string
  preset_type: PresetType
  options: ScanOptions
  parameter_string?: string  // 命令行参数字符串(与BurpSuite兼容)
  is_active: boolean
  created_at?: string
  updated_at?: string
  last_used_at?: string
  use_count: number
}

// 创建预设请求
export interface ScanPresetCreate {
  name: string
  description?: string
  preset_type?: PresetType
  options?: ScanOptions
  parameter_string?: string  // 命令行参数字符串
}

// 更新预设请求
export interface ScanPresetUpdate {
  name?: string
  description?: string
  options?: ScanOptions
  parameter_string?: string  // 命令行参数字符串
  is_active?: boolean
}

// 预设列表响应
export interface ScanPresetListResponse {
  presets: ScanPreset[]
  total: number
  default_preset?: ScanPreset
}

// 配置选项响应（用于下拉菜单）
export interface ConfigOptionsResponse {
  default: ScanPreset | null
  presets: ScanPreset[]
  history: ScanPreset[]
}

// 预设显示选项（用于下拉菜单）
export interface PresetOption {
  label: string
  value: number | 'default' | 'separator'
  preset?: ScanPreset
  type: PresetType | 'separator'
  disabled?: boolean
}

// 默认扫描选项
export const DEFAULT_SCAN_OPTIONS: ScanOptions = {
  level: 1,
  risk: 1,
  technique: 'BEUSTQ',
  timeSec: 5,
  timeout: 30,
  retries: 3,
  delay: 0,
  threads: 1,
  batch: true,
  verbose: 1
}

// 常用DBMS选项
export const DBMS_OPTIONS = [
  { label: '自动检测', value: '' },
  { label: 'MySQL', value: 'MySQL' },
  { label: 'PostgreSQL', value: 'PostgreSQL' },
  { label: 'Microsoft SQL Server', value: 'Microsoft SQL Server' },
  { label: 'Oracle', value: 'Oracle' },
  { label: 'SQLite', value: 'SQLite' },
  { label: 'MariaDB', value: 'MariaDB' },
  { label: 'IBM DB2', value: 'IBM DB2' },
  { label: 'SAP MaxDB', value: 'SAP MaxDB' },
  { label: 'Firebird', value: 'Firebird' },
  { label: 'Sybase', value: 'Sybase' },
  { label: 'H2', value: 'H2' },
  { label: 'HSQLDB', value: 'HSQLDB' },
  { label: 'Informix', value: 'Informix' },
  { label: 'Apache Derby', value: 'Apache Derby' }
]

// 注入技术选项
export const TECHNIQUE_OPTIONS = [
  { label: 'BEUSTQ (全部)', value: 'BEUSTQ' },
  { label: 'B (布尔盲注)', value: 'B' },
  { label: 'E (报错注入)', value: 'E' },
  { label: 'U (联合查询)', value: 'U' },
  { label: 'S (堆叠查询)', value: 'S' },
  { label: 'T (时间盲注)', value: 'T' },
  { label: 'Q (内联查询)', value: 'Q' },
  { label: 'BEU (常用)', value: 'BEU' },
  { label: 'BT (盲注)', value: 'BT' }
]

// Level选项
export const LEVEL_OPTIONS = [
  { label: '1 (快速)', value: 1 },
  { label: '2 (默认)', value: 2 },
  { label: '3 (中等)', value: 3 },
  { label: '4 (完整)', value: 4 },
  { label: '5 (最深入)', value: 5 }
]

// Risk选项
export const RISK_OPTIONS = [
  { label: '1 (安全)', value: 1 },
  { label: '2 (中等)', value: 2 },
  { label: '3 (高风险)', value: 3 }
]
